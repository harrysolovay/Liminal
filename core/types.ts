import type { Expand } from "../util/mod.ts"
import { type AnyType, Type } from "./Type.ts"

export const boolean: Type<boolean> = Type({
  getAtom: () => boolean,
})

export const number: Type<number> = Type({
  getAtom: () => number,
})

export const string: Type<string> = Type({
  getAtom: () => string,
})

export { const_ as const }
export function const_<V extends number | string>(value: V): Type<V> {
  return Type({
    factory: const_,
    args: [value],
  })
}
Object.defineProperty(const_, "name", { value: "const" })

export function array<E extends AnyType>(Element: E): Type<Array<E["T"]>, E["P"]> {
  return Type({
    factory: array,
    args: [Element],
  })
}

export function object<F extends Record<number | string, AnyType>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }, F[keyof F]["P"]> {
  return Type({
    factory: object,
    args: [fields],
  })
}

export function option<X extends AnyType>(Some: X): Type<X["T"] | undefined, X["P"]> {
  return Type({
    factory: option,
    args: [Some],
  })
}

export { enum_ as enum }
function enum_<K extends string>(...members: Array<K>): Type<K> {
  return Type({
    factory: enum_,
    args: members,
  })
}
Object.defineProperty(enum_, "name", { value: "enum" })

export function taggedUnion<
  K extends number | string,
  M extends Record<number | string, AnyType | undefined>,
>(
  tagKey: K,
  members: M,
): Type<
  {
    [V in keyof M]: Expand<({ [_ in K]: V } & (M[V] extends AnyType ? { value: M[V]["T"] } : {}))>
  }[keyof M],
  Extract<M[keyof M], AnyType>["P"]
> {
  return Type({
    factory: taggedUnion,
    args: [tagKey, members],
  })
}

export function transform<F, P extends keyof any, T>(
  From: Type<F, P>,
  f: (value: F) => T,
): Type<T, P> {
  return Type({
    factory: transform,
    args: [From, f],
  })
}

export function deferred<T, P extends keyof any>(getType: () => Type<T, P>): Type<T, P> {
  return Type({
    factory: deferred,
    args: [getType],
  })
}
