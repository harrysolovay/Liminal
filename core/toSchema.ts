import { recombine } from "../util/mod.ts"
import type { Args } from "./Context.ts"
import type { Schema } from "./Schema.ts"
import * as T from "./T.ts"
import { type AnyType, type Type, typeKey } from "./Type.ts"
import { TypeVisitor } from "./TypeVisitor.ts"

export function toSchema<T>(type: Type<T>): Schema {
  return visitor.visit({
    args: {},
    visited: new WeakMap(),
  }, type)
}

const visitor = new TypeVisitor<{
  args: Args
  visited: WeakMap<AnyType, Schema>
}, Schema>()
  .middleware((next, ctx, type, ...factoryArgs) => {
    if (ctx.visited.has(type)) {
      return ctx.visited.get(type)!
    }
    const args = { ...ctx.args }
    const segments: Array<string> = []
    for (const part of type[typeKey].context.parts) {
      if (part.template) {
        segments.unshift(
          recombine(part.template, part.params.map((paramKey) => ctx.args[paramKey])),
        )
      } else {
        Object.assign(args, part.args)
      }
    }
    const description = segments.length ? segments.join(" ") : undefined
    const schema = {
      description,
      ...next(
        {
          args,
          visited: ctx.visited,
        },
        type,
        ...factoryArgs,
      ),
    }
    ctx.visited.set(type, schema)
    return schema
  })
  .add(T.boolean, () => {
    return { type: "boolean" }
  })
  .add(T.Integer, () => ({ type: "integer" }))
  .add(T.number, () => ({ type: "number" }))
  .add(T.string, () => ({ type: "string" }))
  .add(T.array, (ctx, _1, element): Schema => ({
    type: "array",
    items: visitor.visit(ctx, element),
  }))
  .add(T.object, (ctx, _0, fields): Schema => {
    const keys = Object.keys(fields)
    return {
      type: "object",
      properties: Object.fromEntries(
        keys.map((k) => [k, visitor.visit(ctx, fields[k]!)]),
      ),
      additionalProperties: false,
      required: keys,
    }
  })
  .add(T.option, (ctx, _1, Some): Schema => ({
    anyOf: [{ type: "null" }, visitor.visit(ctx, Some)],
  }))
  .add(T.enum, (_0, _1, ...members) => ({
    type: "string",
    enum: members,
  }))
  .add(T.taggedUnion, (ctx, _1, tagKey, members): Schema => {
    return {
      discriminator: tagKey,
      anyOf: Object.entries(members).map(([k, v]) => ({
        type: "object",
        properties: {
          [tagKey]: {
            type: "string",
            const: k,
          },
          ...(v === undefined ? {} : { value: visitor.visit(ctx, v) }),
        },
        required: [tagKey, ...v === undefined ? [] : ["value"]],
        additionalProperties: false,
      })),
    }
  })
  .add(T.transform, (ctx, _0, from): Schema => visitor.visit(ctx, from))
  .add(T.deferred, (ctx, _0, getType): Schema => visitor.visit(ctx, getType()))
