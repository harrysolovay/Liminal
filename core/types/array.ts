import { type AnyType, Type } from "../Type.ts"

export function array<E extends AnyType>(element: E): Type<Array<E["T"]>, E["P"]> {
  return Type({
    kind: "array",
    factory: array,
    args: [element],
    argsLookup: { element },
  })
}
