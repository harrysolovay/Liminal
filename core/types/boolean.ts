import { Type } from "../Type.ts"

export const boolean: Type<boolean> = Type({
  name: "boolean",
  source: {
    getType: () => boolean,
  },
})
