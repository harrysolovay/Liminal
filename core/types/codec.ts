import { Type } from "../Type.ts"

export function codec<T, E, T2>(
  from: Type<T, E>,
  decode: (value: T) => T2,
  encode: (value: T2) => T,
): Type<T2, E> {
  return Type({
    type: "codec",
    self() {
      return codec
    },
    args: [from, decode, encode],
  })
}
