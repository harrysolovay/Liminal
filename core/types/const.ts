import { Type } from "../Type.ts"

export { const_ as const }
export function const_<V extends number | string>(value: V): Type<V> {
  return Type({
    name: "const",
    factory: const_,
    args: [value],
    argsLookup: { value },
  })
}
Object.defineProperty(const_, "name", { value: "const" })
