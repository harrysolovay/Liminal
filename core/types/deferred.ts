import { type AnyType, Type } from "../Type.ts"

export function deferred<X extends AnyType>(get: () => X): X {
  return Type({
    type: "deferred",
    self() {
      return deferred
    },
    args: [get],
  }) as never
}
