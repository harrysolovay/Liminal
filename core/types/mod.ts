export * from "./array.ts"
export * from "./boolean.ts"
export * from "./enum.ts"
export * from "./number.ts"
export * from "./object.ts"
export * from "./option.ts"
export * from "./string.ts"
export * from "./taggedUnion.ts"
export * from "./transform.ts"
