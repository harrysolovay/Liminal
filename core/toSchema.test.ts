import { assertSnapshot } from "@std/testing/snapshot"
import { toSchema } from "./toSchema.ts"
import type { Type } from "./Type.ts"
import * as T from "./types.ts"

Deno.test("toSchema", async (t) => {
  await t.step("boolean", async (t) => {
    await assertTypeSnapshot(t, T.boolean)
  })

  await t.step("number", async (t) => {
    await assertTypeSnapshot(t, T.number)
  })

  await t.step("string", async (t) => {
    await assertTypeSnapshot(t, T.string)
  })

  await t.step("array", async (t) => {
    await assertTypeSnapshot(t, T.array(T.string))
  })

  await t.step("object", async (t) => {
    await assertTypeSnapshot(
      t,
      T.object({
        a: T.boolean,
        b: T.number,
        c: T.string,
      }),
    )
  })

  await t.step("option", async (t) => {
    await assertTypeSnapshot(t, T.option(T.string))
  })

  await t.step("enum", async (t) => {
    await assertTypeSnapshot(t, T.enum("A", "B", "C"))
  })

  await t.step("tagged union", async (t) => {
    await assertTypeSnapshot(
      t,
      T.taggedUnion("type", {
        A: T.boolean,
        B: T.number,
        C: T.string,
      }),
    )
  })
})

async function assertTypeSnapshot(t: Deno.TestContext, value: Type<any>): Promise<void> {
  await assertSnapshot(t, toSchema(value))
  const withContext = value`One.`
  await assertSnapshot(t, toSchema(withContext))
  await assertSnapshot(t, toSchema(withContext`Two.`))
}
