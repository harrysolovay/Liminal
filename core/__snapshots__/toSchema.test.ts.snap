export const snapshot = {};

snapshot[`toSchema > boolean 1`] = `
{
  "\$defs": {
    "0": {
      description: undefined,
      type: "boolean",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > boolean 2`] = `
{
  "\$defs": {
    "0": {
      description: "One.",
      type: "boolean",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > boolean 3`] = `
{
  "\$defs": {
    "0": {
      description: "One. Two.",
      type: "boolean",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > number 1`] = `
{
  "\$defs": {
    "0": {
      description: undefined,
      type: "number",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > number 2`] = `
{
  "\$defs": {
    "0": {
      description: "One.",
      type: "number",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > number 3`] = `
{
  "\$defs": {
    "0": {
      description: "One. Two.",
      type: "number",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > integer 1`] = `
{
  "\$defs": {
    "0": {
      description: undefined,
      type: "integer",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > integer 2`] = `
{
  "\$defs": {
    "0": {
      description: "One.",
      type: "integer",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > integer 3`] = `
{
  "\$defs": {
    "0": {
      description: "One. Two.",
      type: "integer",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > string 1`] = `
{
  "\$defs": {
    "0": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > string 2`] = `
{
  "\$defs": {
    "0": {
      description: "One.",
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > string 3`] = `
{
  "\$defs": {
    "0": {
      description: "One. Two.",
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > array 1`] = `
{
  "\$defs": {
    "0": {
      description: undefined,
      items: {
        "\$ref": "#/\$defs/1",
      },
      type: "array",
    },
    "1": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > array 2`] = `
{
  "\$defs": {
    "0": {
      description: "One.",
      items: {
        "\$ref": "#/\$defs/1",
      },
      type: "array",
    },
    "1": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > array 3`] = `
{
  "\$defs": {
    "0": {
      description: "One. Two.",
      items: {
        "\$ref": "#/\$defs/1",
      },
      type: "array",
    },
    "1": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > object 1`] = `
{
  "\$defs": {
    "0": {
      additionalProperties: false,
      description: undefined,
      properties: {
        a: {
          "\$ref": "#/\$defs/1",
        },
        b: {
          "\$ref": "#/\$defs/2",
        },
        c: {
          "\$ref": "#/\$defs/3",
        },
      },
      required: [
        "a",
        "b",
        "c",
      ],
      type: "object",
    },
    "1": {
      description: undefined,
      type: "boolean",
    },
    "2": {
      description: undefined,
      type: "number",
    },
    "3": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > object 2`] = `
{
  "\$defs": {
    "0": {
      additionalProperties: false,
      description: "One.",
      properties: {
        a: {
          "\$ref": "#/\$defs/1",
        },
        b: {
          "\$ref": "#/\$defs/2",
        },
        c: {
          "\$ref": "#/\$defs/3",
        },
      },
      required: [
        "a",
        "b",
        "c",
      ],
      type: "object",
    },
    "1": {
      description: undefined,
      type: "boolean",
    },
    "2": {
      description: undefined,
      type: "number",
    },
    "3": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > object 3`] = `
{
  "\$defs": {
    "0": {
      additionalProperties: false,
      description: "One. Two.",
      properties: {
        a: {
          "\$ref": "#/\$defs/1",
        },
        b: {
          "\$ref": "#/\$defs/2",
        },
        c: {
          "\$ref": "#/\$defs/3",
        },
      },
      required: [
        "a",
        "b",
        "c",
      ],
      type: "object",
    },
    "1": {
      description: undefined,
      type: "boolean",
    },
    "2": {
      description: undefined,
      type: "number",
    },
    "3": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > option 1`] = `
{
  "\$defs": {
    "0": {
      anyOf: [
        {
          type: "null",
        },
        {
          "\$ref": "#/\$defs/1",
        },
      ],
      description: undefined,
    },
    "1": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > option 2`] = `
{
  "\$defs": {
    "0": {
      anyOf: [
        {
          type: "null",
        },
        {
          "\$ref": "#/\$defs/1",
        },
      ],
      description: "One.",
    },
    "1": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > option 3`] = `
{
  "\$defs": {
    "0": {
      anyOf: [
        {
          type: "null",
        },
        {
          "\$ref": "#/\$defs/1",
        },
      ],
      description: "One. Two.",
    },
    "1": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > enum 1`] = `
{
  "\$defs": {
    "0": {
      description: undefined,
      enum: [
        "A",
        "B",
        "C",
      ],
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > enum 2`] = `
{
  "\$defs": {
    "0": {
      description: "One.",
      enum: [
        "A",
        "B",
        "C",
      ],
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > enum 3`] = `
{
  "\$defs": {
    "0": {
      description: "One. Two.",
      enum: [
        "A",
        "B",
        "C",
      ],
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > tagged union 1`] = `
{
  "\$defs": {
    "0": {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "A",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/1",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "B",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/2",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "C",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/3",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: undefined,
      discriminator: "type",
    },
    "1": {
      description: undefined,
      type: "boolean",
    },
    "2": {
      description: undefined,
      type: "number",
    },
    "3": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > tagged union 2`] = `
{
  "\$defs": {
    "0": {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "A",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/1",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "B",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/2",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "C",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/3",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "One.",
      discriminator: "type",
    },
    "1": {
      description: undefined,
      type: "boolean",
    },
    "2": {
      description: undefined,
      type: "number",
    },
    "3": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`toSchema > tagged union 3`] = `
{
  "\$defs": {
    "0": {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "A",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/1",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "B",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/2",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "C",
              type: "string",
            },
            value: {
              "\$ref": "#/\$defs/3",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "One. Two.",
      discriminator: "type",
    },
    "1": {
      description: undefined,
      type: "boolean",
    },
    "2": {
      description: undefined,
      type: "number",
    },
    "3": {
      description: undefined,
      type: "string",
    },
  },
  additionalProperties: false,
  properties: {
    value: {
      "\$ref": "#/\$defs/0",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;
